### PR TITLE
General python support

### PR DESCRIPTION
**Pull request by VikingScientist**
In awaiting the conclusion from #141, I tried to show how a general version-agnostic approach could look. In practice, you only need to add the line `from __future__ import print_function` to all python files and they will in 99% of the times work for both versions. At least when considering small algorithms such as this project here and not huge complex interconnected libraries that have lots of dependencies.

### CODE REVIEW

**Comment by jiegillet** [comment]
I think the consensus is that we don't want to have a compromise on versions. Getting rid of `py2` and `py3` and use `py` is great, but I don't think we want to include `from __future__ import print_function ` at all. People who want to use python 3 code using python 2 should add that themselves.

File: chapters/euclidean_algorithm/euclidean.md [hunk 1]
```diff
@@ -89,7 +89,7 @@ Program.cs
 [import, lang="javascript"](code/javascript/euclidean_example.js)
 {% sample lang="py2" %}
```
Comment (jiegillet) [inline chapters/euclidean_algorithm/euclidean.md]: You missed a py2 here

File: chapters/tree_traversal/tree_traversal.md [hunk 2]
```diff
@@ -41,7 +39,7 @@ Because of this, the most straightforward way to traverse the tree might be recu
 {% sample lang="js" %}
 [import:15-23, lang:"javascript"](code/javascript/Tree_example.js)
 {% sample lang="py2" %}
```
Comment (jiegillet) [inline chapters/tree_traversal/tree_traversal.md]: py 2 here

File: chapters/tree_traversal/tree_traversal.md [hunk 3]
```diff
@@ -41,7 +39,7 @@ Because of this, the most straightforward way to traverse the tree might be recu
...
-[import:8-16, lang:"python"](code/python2/Tree_example.py)
+[import:7-15, lang:"python"](code/python/Tree_example.py)
 {% sample lang="py3" %}
```
Comment (jiegillet) [inline chapters/tree_traversal/tree_traversal.md]: py3 here (although there is no code)

**Comment by leios** [comment]
As a quick note the `py` label should be defined in the `book.json` file in the root directory. this will make sure that all `py` labels will have the correct tag at the top of the screen ("Python" instead of "Python 2" or "Python 3"). I can do this and submit a PR to your branch @VikingScientist , or you can change 

```
                                {
                                        "lang": "py2",
                                        "name": "Python 2"
                                },
                                {
                                        "lang": "py3",
                                        "name": "Python 3"
                                },

```

to

```
                                {
                                        "lang": "py",
                                        "name": "Python"
                                },

```

and then just check to make sure the language is correct and syntax highlighting is right by running `gitbook serve`.

**Comment by jiegillet** [comment]
@leios is there a place we could mention explicitly our stance on python version?

**Comment by leios** [comment]
@jiegillet I think I we already did in #141. Maybe we should make language-specific style guides and mention it in here?

**Comment by jiegillet** [comment]
No I mean in the book. The topic is bound to come up. It should be stated somewhere, maybe?

**Comment by leios** [comment]
Yeah, that's what I meant too. We should add some style guidelines to the *How to Contribute* section and for python, we just say that we are sticking to python 3 for this.

I am not sure where else to put it, but @julianschacher was mentioning he wanted to create a *How to Contribute* guide... So that would be a good place too. 

Maybe we should move this to a separate issue?

**Comment by jiegillet** [comment]
No #141 is the right place to discuss that. This is slightly wrong though, so let's get back to code reviewing, sorry @VikingScientist

**Comment by VikingScientist** [comment]
Hehe. This is fine. I initiated the discussion, and if it is discussed here or in #141 really doesn't bother me. I agree with @leios though: we should put the "python 3 only"-stance somewhere (i.e. in the How to contribute section) so that it is available for new contributers. The issue tracker will fade; in fact auto-close after this PR is merged, so it's not a good long-term documentation.

There's a few upcoming commits more where I make the requested changes to `book.json` as well as fixing the last python-code in the FFT. After this, all version-specific code is gone, so I'll remove py2 and py3 from `book.json`.

**Comment by jiegillet** [comment]
Coolz. Ping me when you feel it's ready for a merge.

**Comment by leios** [comment]
Oh yeah, if this PR removes all version specific stuff from the whole book, it should be fine to remove `py2` and `py3`. I just thought there were more python implementations hiding somewhere.

**Comment by VikingScientist** [comment]
I did `grep` for any mention of py2 or py3, but can't find any trace of it anymore.

I like fixup commits such as 2356130896210f6f3ebe69c6d8a38407f4902c31 as it makes the reviewers life easier, but I want these to be squashed with it's referencing commit before being merged into trunk. So if I get a thumbs up on this so far then I'll run `git rebase -i HEAD~6 --autosquash` and then it's ready to be merged.

**Comment by Butt4cak3** [comment]
We always squash and merge PRs (through the GH interface) anyway. There's no need to manually squash. But thanks for the consideration!

**Comment by VikingScientist** [comment]
Rebased to resolve some merge conflicts. Ready to be looked at @jiegillet

**Comment by leios** [comment]
Great!, so once this is merged, we can have a look at the python 3 implementations #146 #144 #138 , all of which should have modifications in their markdown files. 

I am mentioning this now so we remember on those later.

**Pull request merged**